File: src/global_planner.cpp
#include <ros/ros.h>
#include <queue>

#include <grid_map_ros/grid_map_ros.hpp>
#include <priority_point.h>
#include <global_planner.h>
#include <vector>
#include <math.h>

//////////////////////////////////////////////////////////////////////////////
//////////////////////////// Theta* //////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////
// ????????????
PriorityPoint neighbors[8];
PriorityPoint current;
// double cost_so_far[300][200];
// int come_from[300][200][2];
// ????????????


std::vector<grid_map::Index> makePath_ThetaStar(PriorityPoint start, PriorityPoint goal, grid_map::GridMap &grid_map)
{
  std::string layer = "static";
  std::priority_queue<PriorityPoint, std::vector<PriorityPoint>, myCompare> openSet;
  size_t max_x = grid_map.getSize()[0];
  size_t max_y = grid_map.getSize()[1];
  double *cost_so_far = new double[max_x*max_y];
  int *come_from = new int[max_x*max_y*2];

  std::fill_n(cost_so_far, max_x * max_y, std::numeric_limits<double>::max());
  cost_so_far[start.x*max_y+start.y] = 0;
  come_from[start.x*max_y*2+start.y*2+0] = start.x;
  come_from[start.x*max_y*2+start.y*2+1] = start.y;
  openSet.push(start);
  // grid_cost=5;
  while (!openSet.empty())
  {
    current = openSet.top();
    openSet.pop();
    if (current == goal)
    {
      break;
    }
    current.GetNeighbors(neighbors);
    double current_cost = cost_so_far[current.x*max_y+current.y];
    // TODO: problem is somewhere here 
    int parent_carent[2] ={come_from[current.x*max_y*2+current.y*2+0], come_from[current.x*max_y*2+current.y*2+1]};
    for (int i = 0; i < 8; i++)
    {
      // TODO: put there into GetNeighbors()
      if (!neighbors[i].OnMap(max_x, max_y))
      {
        continue;
      }
      bool onLine = lineOfSight(parent_carent[0], parent_carent[1], neighbors[i].x, neighbors[i].y, grid_map);
      if (onLine)
      {
        double new_cost = cost_so_far[parent_carent[0]*max_y+parent_carent[1]] + HeuristicEvclid(parent_carent, neighbors[i]);
        if (new_cost < cost_so_far[neighbors[i].x*max_y+neighbors[i].y])
        {
          cost_so_far[neighbors[i].x*max_y+neighbors[i].y] = new_cost;
	        neighbors[i].priority = HeuristicEvclid(neighbors[i], goal) + new_cost;
          openSet.push(neighbors[i]);
          come_from[neighbors[i].x*max_y*2+neighbors[i].y*2+0] = parent_carent[0];
          come_from[neighbors[i].x*max_y*2+neighbors[i].y*2+1] = parent_carent[1];
        }
      }
      else
      {
        double neighbor_price = grid_map.at(layer, grid_map::Index({neighbors[i].x, neighbors[i].y})) + neighbors[i].priority;
        double new_cost = current_cost + neighbor_price;
        if (new_cost < cost_so_far[neighbors[i].x*max_y+neighbors[i].y])
        {
          cost_so_far[neighbors[i].x*max_y+neighbors[i].y] = new_cost;
          neighbors[i].priority =HeuristicEvclid(neighbors[i], goal) + new_cost;
          openSet.push(neighbors[i]);
          come_from[neighbors[i].x*max_y*2+neighbors[i].y*2+0] = current.x;
          come_from[neighbors[i].x*max_y*2+neighbors[i].y*2+1] = current.y;
        }
      }
    }
  }

  // path.clear();
  std::vector<grid_map::Index> path;
  PriorityPoint temp_point;
  
  while (current != start)
  {
    path.push_back({current.x, current.y});
    temp_point.x = come_from[current.x*max_y*2+current.y*2+0];
    temp_point.y = come_from[current.x*max_y*2+current.y*2+1];
    printf("%d %d",temp_point.x,temp_point.y);
    // ROS_INFO(std::to_string(temp_point.x));
    // ROS_INFO(std::to_string(temp_point.y));
    current = temp_point;
  }
  path.push_back({current.x, current.y});
  delete[] cost_so_far;
  delete[] come_from;

  return path;
}


bool lineOfSight(int i1, int j1, int i2, int j2, grid_map::GridMap &grid_map)
{
  grid_map::Index p1 = {i1, j1};
  grid_map::Index p2 = {i2, j2};
  std::string layer = "static";
  for (grid_map::LineIterator iterator(grid_map, p1, p2); !iterator.isPastEnd(); ++iterator)
  {
    if (grid_map.at(layer, *iterator))
    {
      return false;
    }
  }
  return true;
}


//////////////////////////////////////////////////////////////////////////////
////////////////////////////// support ///////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////
bool isNear(PriorityPoint &p1,PriorityPoint &p2,unsigned int dist)
{
  unsigned int a = abs(p1.x - p2.x);
  unsigned int b = abs(p1.y - p2.y);
  return (a + b)<dist;

}

//////////////////////////////////////////////////////////////////////////////
//////////////////////////////// Heuristics //////////////////////////////////
//////////////////////////////////////////////////////////////////////////////
float HeuristicEvclid(const PriorityPoint &source, const PriorityPoint &target)
{
  return sqrt(pow(source.x - target.x, 2) + pow(source.y - target.y, 2));
}

float HeuristicEvclid(int* source, const PriorityPoint &target)
{
  return sqrt(pow(source[0] - target.x, 2) + pow(source[1] - target.y, 2));
}


unsigned int HeuristicChebishev(const PriorityPoint &source, const PriorityPoint &target)
{
  unsigned int a = abs(source.x - target.x);
  unsigned int b = abs(source.y - target.y);
  return a > b ? a  : b ;
}
unsigned int HeuristicManhetten(const PriorityPoint &source, const PriorityPoint &target)
{
  return (abs(source.x - target.x) + abs(source.y - target.y));
}
unsigned int HeuristicManhetten(int* source, const PriorityPoint &target)
{
  return (abs(source[0] - target.x) + abs(source[1] - target.y));
}

File: scripts/Utils/GlobalPlanner.py
import math
import numpy as np



def select_next_goal(state, current_goal, path,radius = 0.05):
    for pose in range(len(path)-1):
        dist = math.sqrt(math.pow(state[0] - path[pose,0], 2) + math.pow(state[1] - path[pose,1], 2))
        if dist<radius:
            return path[pose+1]
    return current_goal

class GlobalPlanner:

    def __init__(self):
        pass

    def get_nearest_goal(self, path, position):
        pass

    def get_neighbors(self, position, grid):
        # add neighbors
        neighbors =np.array([
            [position[0]-1, position[1]],
            [position[0]-1, position[1]-1],
            [position[0], position[1]-1],

            [position[0]+1, position[1]],
            [position[0]+1, position[1]+1],
            [position[0], position[1]+1],

            [position[0]-1, position[1]+1],
            [position[0]+1, position[1]-1],
        ])
        # remove out of map
        p=0
        for n in range(len(neighbors)):
            if (neighbors[-1-p] <= 0).any() or (neighbors[-1-p] >= grid.shape).any():
                neighbors = np.delete(neighbors,-1-p,0) #neighbors.pop([-1-n])
                continue
            p+=1
        return neighbors

    def line_Of_sight(self, point1, point2, grid):
        # n = 4 ?
        # n = 4
        # dxy = (np.sqrt((point1[0] - point2[0]) **
        #                2 + (point1[1] - point2[1]) ** 2)) * n
        # i = np.rint(np.linspace(point1[0], point2[0], dxy)).astype(int)
        # j = np.rint(np.linspace(point1[1], point2[1], dxy)).astype(int)
        # dxy = max(abs(point1[0]-point2[0])+1,abs(point1[1]-point2[1])+1)
        dxy = self.heuristic_manhetten(point1,point2)
        i = np.linspace(point1[0], point2[0],dxy).astype(int)
        j = np.linspace(point1[1], point2[1],dxy).astype(int)
        has_collision = np.any(grid[i, j])
        return has_collision

    def heuristic_evclid(self, source, target):
        return math.sqrt(math.pow(source[0] - target[0], 2) + math.pow(source[1] - target[1], 2))

    def heuristic_chebishev(self, source, target):
        a = abs(source[0] - target[0])
        b = abs(source[1] - target[1])
        if b > a:
            a = b
        return a

    def heuristic_manhetten(self, source, target):
        return (abs(source[0] - target[0]) + abs(source[1] - target[1]))


class GlobalPlannerThetaStar(GlobalPlanner):
    def __init__(self):
        pass

    def calc_path(self, start, goal, grid):
        # grid = grid.astype(bool)
        open_set = [start]
        cost_so_far = {hash(start.tostring()): 0}
        came_from = {hash(start.tostring()): start}
        while len(open_set) > 0:
            current = open_set.pop()  # check
            parent = came_from[hash(current.tostring())]
            if (current == goal).all():
                break
            neighbors = self.get_neighbors(current, grid)
            current_cost = cost_so_far[hash(current.tostring())]
            parent_cost = cost_so_far[hash(parent.tostring())]
            for neighbor in neighbors:
                if hash(neighbor.tostring()) not in cost_so_far:
                    cost_so_far[hash(neighbor.tostring())] = 999999999999
                neighbor_cost = cost_so_far[hash(neighbor.tostring())]
                has_collision = self.line_Of_sight(parent, neighbor, grid)
                if has_collision:
                    new_cost = current_cost + \
                        self.heuristic_evclid(current, neighbor)
                    if new_cost < neighbor_cost:
                        cost_so_far[hash(neighbor.tostring())] = new_cost
                        came_from[hash(neighbor.tostring())] = current
                        open_set.append(neighbor)
                else:
                    new_cost = parent_cost + \
                        self.heuristic_evclid(parent, neighbor)
                    if new_cost < neighbor_cost:
                        cost_so_far[hash(neighbor.tostring())] = new_cost
                        came_from[hash(neighbor.tostring())] = parent
                        open_set.append(neighbor)
        current = goal
        path = [current]
        try:
            came_from[hash(current.tostring())]
        except:
            path.append(start)
            return np.array(path)
        while (current != start).all():
            current = came_from[hash(current.tostring())]
            path.append(current)
        return np.array(path)


    # def calc_path(self, start, goal, grid):
    #     # grid = grid.astype(bool)
    #     open_set = [start]
    #     cost_so_far = {hash(start.tostring()): 0}
    #     came_from = {start: start}
    #     while len(open_set) > 0:
    #         current = open_set.pop()  # check
    #         parent = came_from[current]
    #         if current == goal:
    #             break
    #         neighbors = self.get_neighbors(current, grid)
    #         current_cost = cost_so_far[current]
    #         parent_cost = cost_so_far[parent]
    #         for neighbor in neighbors:
    #             neighbor_cost = cost_so_far[neighbor]
    #             has_collision = self.line_Of_sight(parent, neighbor, grid)
    #             if has_collision:
    #                 new_cost = current_cost + \
    #                     self.heuristic_evclid(current, neighbor)
    #                 if new_cost < neighbor_cost:
    #                     cost_so_far[neighbor] = new_cost
    #                     come_from[neighbor] = current
    #                     open_set.append(neighbor)
    #             else:
    #                 new_cost = parent_cost + \
    #                     self.heuristic_evclid(parent, neighbor)
    #                 if new_cost < neighbor_cost:
    #                     cost_so_far[neighbor] = new_cost
    #                     come_from[neighbor] = parent
    #                     open_set.append(neighbor)
    #     current = goal
    #     path = [current]
    #     if came_from[current] is None:
    #         return path
    #     while current != start:
    #         current = came_from[current]
    #         path.append(current)
    #     return path


# planner = GlobalPlannerThetaStar()
# exit()
